From d6d3dcb65f3288a8533f9a91eb2eb58862297f1b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Florian=20M=C3=BCllner?= <fmuellner@redhat.com>
Date: Mon, 23 Apr 2012 19:12:19 +0200
Subject: [PATCH] window: Do not open password dialogs underneath fullscreen
 windows

We consider password dialogs important enough that we do not want
them to be mapped underneath a fullscreen window, so add an exception
to focus stealing prevention for that case.
---
 src/core/window.c |   27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/src/core/window.c b/src/core/window.c
index 2f2f800..5ed0892 100644
--- a/src/core/window.c
+++ b/src/core/window.c
@@ -1971,6 +1971,20 @@ __window_is_terminal (MetaWindow *window)
   return FALSE;
 }
 
+static gboolean
+__window_is_password_dialog (MetaWindow *window)
+{
+  if (window == NULL || window->res_class == NULL)
+    return FALSE;
+
+  if (strcmp (window->res_class, "Polkit-gnome-authentication-agent-1") == 0)
+    return TRUE;
+  else if (strcmp (window->res_class, "Gcr-prompter") == 0)
+    return TRUE;
+
+  return FALSE;
+}
+
 /* This function determines what state the window should have assuming that it
  * and the focus_window have no relation
  */
@@ -1995,6 +2009,19 @@ window_state_on_map (MetaWindow *window,
       return;
     }
 
+  /* We consider password dialogs important enough that we do not
+   * want them to end up hidden underneath a fullscreen window due
+   * to focus-stealing prevention.
+   */
+  if (!*takes_focus &&
+      window->display->focus_window &&
+      window->display->focus_window->fullscreen &&
+      __window_is_password_dialog (window))
+    {
+      *takes_focus = TRUE;
+      return;
+    }
+
   /* Terminal usage may be different; some users intend to launch
    * many apps in quick succession or to just view things in the new
    * window while still interacting with the terminal.  In that case,
-- 
1.7.10

